[Vis: Default editor] EUIficate Sub agg control

maryia-lapata opened the pull request:
## Summary

EUIfication of Sub agg control for aggregation parameter in Default Editor, Data tab.
Part of #30922.

![image](https://user-images.githubusercontent.com/11524637/58882478-0457b380-86e5-11e9-8abd-9de926a7fc8d.png)

_Steps to reproduce_: create a visualization, choose an aggregation from `Parent Pipeline Aggregations` group (e.g. `Derivative`).

#### Details

- `validate-agg` directive was removed
- [`agg_select.js`](https://github.com/elastic/kibana/blob/f492e666d5cbe66c4fc8b39ae004db4a31670107/src/legacy/ui/public/vis/editors/default/agg_select.js) was updated. Now it has the same validation flow as `agg_param.js` has
- functions `isCompatibleAgg` and `safeMakeLabel` were moved to a separate utils file ([`agg_utils.ts`](https://github.com/elastic/kibana/blob/55d4306f0ad939eac5c143723bc0c71555c4eb62/src/legacy/ui/public/agg_types/agg_utils.ts))

#### Screenshot:

Please pay your attention that validation a bit changed: 

_before_: when a user chooses incompatible bucket aggregation, **metrics** aggregation input was in red
_after_: when a user chooses incompatible bucket aggregation, **bucket** aggregation input is in red

I changed this validation because it's a bit difficult to implement _before_ case until `agg_params.js` will be migrated. Also I find it logical to make **bucket** aggregation input in red as that agg is incompatible. Reviewers, please, let me know if it's applicable.

Also now the error message is displayed underneath the agg form control:

![Jun-05-2019 14-32-18](https://user-images.githubusercontent.com/11524637/58953367-ee0e2e00-879e-11e9-8d1a-b34b31364736.gif)



### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~~[ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- ~~[ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- ~~[ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
- ~~[ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/17115/)

cchaos commented:
It doesn't seem to automatically put the sub-agg in an error state **until** you interact with it. Is there any way to force that state before you interact with it?

<img src="https://d.pr/free/i/VOBpTz+" />

Also is there any way to change this message so that it is just help text underneath the sub-agg form control?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/161/)

maryia-lapata commented:
>It doesn't seem to automatically put the sub-agg in an error state until you interact with it. Is there any way to force that state before you interact with it?

Oh yes, I updated the PR. Thank you for noticing.

> Also is there any way to change this message so that it is just help text underneath the sub-agg form control?

I moved this message into agg selector's error section. 
I updated a screenshot in the description, please take a look.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/182/)

elasticmachine commented:
Pinging @elastic/kibana-app

maryia-lapata commented on src/legacy/ui/public/agg_types/metrics/lib/parent_pipeline_agg_controller.js:
```diff
@@ -79,9 +70,6 @@ const parentPipelineAggController = function ($scope) {
 
     // we aren't creating a custom aggConfig
     if (metricAgg !== 'custom') {
-      if (!$scope.state.aggs.find(agg => agg.id === metricAgg)) {
```
This logic was moved to `MetricAggParamEditor` component: https://github.com/elastic/kibana/pull/37979/files?file-filters%5B%5D=.tsx&owned-by%5B%5D=#diff-d09c7d350e3017c52e2e151fba97ef27R54

maryia-lapata commented on src/legacy/ui/public/vis/editors/default/agg_params.html:
```diff
@@ -13,12 +13,6 @@
 
-<div ng-if="agg.error" class="form-group">
-  <p class="visEditorAggParam__error ng-binding">
-    {{agg.error}}
```
Now this error message is displayed in aggregation selector control: https://github.com/elastic/kibana/pull/37979/files?file-filters%5B%5D=.tsx&owned-by%5B%5D=#diff-9c2f5870ef0ffd3edbde4d2072e86430R98

cchaos replied:
Can you also change the exclamation point of the error message to just a period:

<img src="https://d.pr/free/i/xjHl8Y+" width="50%" />

maryia-lapata replied:
Done.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/490/)

maryia-lapata commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/504/)

sulemanof commented on src/legacy/ui/public/agg_types/buckets/_terms_helper.tsx:
```diff
@@ -35,19 +32,4 @@ const aggFilter = [
-}
-
-export { aggFilter, isCompatibleAgg, safeMakeLabel };
+export { aggFilter };
```
There is no necessity to store it in a separate file now. Please, move it to `order_agg.tsx`

sulemanof commented on src/legacy/ui/public/agg_types/controls/sub_agg.html:
```diff
@@ -1,43 +1,10 @@
-    </ng-form>
-  </div>
+<div ng-controller="aggParam.controller" ng-show="agg.params.metricAgg === 'custom'" class="visEditorAgg__subAgg">
+  <ng-form name="customMetricForm">
```
Please, remove extra `ng-form`

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/524/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/528/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/879/)

wylieconlon commented:
I'm a bit confused about the warnings here. In both this branch and on master, I see the blue arrow indicating that a Derivative is valid without any X axis: <img width="333" alt="Screenshot 2019-06-07 13 20 21" src="https://user-images.githubusercontent.com/666475/59121765-03ae5f80-8927-11e9-84ca-75116d847f0a.png">

I see that this is not a regression, but it was confusing and seems related. What do you think?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1158/)

maryia-lapata commented:
> I'm a bit confused about the warnings here. In both this branch and on master, I see the blue arrow indicating that a Derivative is valid without any X axis: <img alt="Screenshot 2019-06-07 13 20 21" width="333" src="https://user-images.githubusercontent.com/666475/59121765-03ae5f80-8927-11e9-84ca-75116d847f0a.png">
> 
> I see that this is not a regression, but it was confusing and seems related. What do you think?


@wylieconlon I tested on 6.7, the issue is reproduced. So for me it seems like a bug. I create an issue: 
https://github.com/elastic/kibana/issues/38537

maryia-lapata commented:
I believe that #38537 doesn't block migration.  @wylieconlon could please continue with review?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1587/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1626/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1633/)
